### PR TITLE
MADScaler

### PR DESCRIPTION
**Pull request by cxhernandez**
- [x] Implement feature / fix bug
 - [x] Add tests
 - [x] Update changelog

re: https://github.com/msmbuilder/msmbuilder/issues/849

### CODE REVIEW

**Comment by cxhernandez** [comment]
The online algorithm (implemented in cython) is still 2 order of magnitude slowdown than a naive numpy implementation:

```python
In [1]: def mad_naive(X):
    Y = np.concatenate(X)
    med = np.median(Y, axis=0)
    mad = np.median(np.abs(Y - med), axis=0)
    return mad, med
   ....:

In [2]: from msmbuilder.preprocessing import MADScaler; mad = MADScaler()

In [23]: %timeit mad.fit(X)
1 loop, best of 3: 814 ms per loop

In [24]: %timeit mad_naive(X)
100 loops, best of 3: 19.1 ms per loop
```

Not sure if the memory savings are entirely worth it....

**Comment by mpharrigan** [comment]
1. Worried about adding more cython (maintenence) esp. if it's slower
2. use `cython -a` to find hotspots
3. Your for loop is probably doing python-y things and ruining your code. Use `cdef int i; for i in range()`
4. use cython's typed memoryviews instead of the verbose numpy syntax

**Comment by mpharrigan** [comment]
5. your use of high-level numpy function within the loop is certainly doing python-y things and ruining your code

**Comment by cxhernandez** [comment]
Thanks for the recs! That got it down to about the same speed as the naive implementation (+/- a few hundred microseconds). 

> Worried about adding more cython (maintenence) esp. if it's slower

It's a pretty small file. I'll do some tests to see if this works better than the `RobustScaler`. Another option is to see if the `scipy` folks are interested and then we won't have to worry about it.

**Comment by mpharrigan** [comment]
> Another option is to see if the scipy folks are interested and then we won't have to worry about it.

You should probably do this regardless :)

File: msmbuilder/preprocessing/_mad.pyx [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+@cython.infer_types(True)
+def mad_estimate(ndarray[np.float64_t, ndim=2] arr,
+                 ndarray[np.float64_t, ndim=1] mad,
+                 ndarray[np.float64_t, ndim=1] med,
+                 double eta):
+    cdef int i
+    cdef int n_rows = arr.shape[0]
+    cdef int n_cols = arr.shape[1]
+    cdef double[:, ::1] arr2 = arr
+    cdef double[::1] mad2 = mad
```
Comment (mpharrigan) [inline msmbuilder/preprocessing/_mad.pyx]: Just change the type in the function signature. It works I promise!

**Comment by mpharrigan** [comment]
cdef `dev1` and `dev2`

You should be able to get it so the loop in `cython -a` has no yellow at all

File: msmbuilder/preprocessing/_mad.pyx (hunk 1 above)
Comment (cxhernandez) [inline msmbuilder/preprocessing/_mad.pyx]: Like this?

```cython
def mad_estimate(double[:, ::1] arr,
                 double[::1] mad,
                 double[::1]] med,
                 double eta):
```

Comment (mpharrigan) [inline msmbuilder/preprocessing/_mad.pyx]: yes

**Comment by cxhernandez** [comment]
> You should probably do this regardless :)

Looking at https://github.com/scikit-learn/scikit-learn/pull/2176#issuecomment-179303371, this seems like a pretty hard task....

**Comment by mpharrigan** [comment]
Yeah it seems pretty stringent. I get where they're coming from---there's value lost when giving *too* many choices. The response to the linked comment was pretty tone deaf

**Pull request merged**